File: Webclient/src/app/stdlib/interval_bucket.ts
import {iterable_filterMap, iterable_map} from "./iterable_higher_order";

class Bucket {
    private content: Map<number, Array<[number, number]>> = new Map();

    constructor(
        public start: number,
        public end: number
    ) {
    }

    insert(value_index: number, start: number, end: number): void {
        if (!this.content.has(value_index))
            this.content.set(value_index, []);
        this.content.get(value_index).push([start, end]); // TODO: Interval merging?
    }

    find(timestamp: number): Array<[number, number, number]> {
        // @ts-ignore
        return iterable_filterMap(this.content.entries(),
            ([index, intervals]) => {
                const result = intervals.find(([start, end]) => start <= timestamp && end >= timestamp);
                if (result !== undefined)
                    return [index, ...result];
                return undefined;
            });
    }

    find_within_range(start: number, end: number): Array<[number, number, number]> {
        if (start <= this.start && end >= this.end) {
            // @ts-ignore
            return iterable_map(this.content.entries(),
                ([value_index, interval]) => [value_index, ...interval]);
        } else if (this.start <= start && end >= this.end) {
            // @ts-ignore
            return iterable_filterMap(this.content.entries(),
                ([value_index, intervals]) => {
                    const result = intervals.find(([i_start, i_end]) => i_end >= start);
                    if (result !== undefined)
                        return [value_index, ...result];
                    return undefined;
                });
        } else if (start <= this.start && this.end >= end) {
            // @ts-ignore
            return iterable_filterMap(this.content.entries(),
                ([value_index, intervals]) => {
                    const result = intervals.find(([i_start, i_end]) => i_start <= end);
                    if (result !== undefined)
                        return [value_index, ...result];
                    return undefined;
                });
        }
        // @ts-ignore
        return iterable_filterMap(this.content.entries(),
            ([value_index, intervals]) => {
                const result = intervals.find(([i_start, i_end]) => i_start <= end && i_end >= start);
                if (result !== undefined)
                    return [value_index, ...result];
                return undefined;
            });
    }
}

export class IntervalBucket {
    private readonly buckets: Array<Bucket>;

    constructor(
        private start: number,
        private end: number,
        private step_size: number
    ) {
        const num_buckets: number = Math.ceil((end - start) / step_size);
        this.buckets = Array(num_buckets);
        for (let i = 0; i < num_buckets; ++i)
            this.buckets[i] = new Bucket(start + i * step_size, start + (i + 1) * step_size);

    }

    insert(start: number, end: number, value_index: number): void {
        start = Math.max(this.start, start);
        end = Math.min(this.end, end);
        const index_start = this.get_bucket_index(start);
        const index_end = this.get_bucket_index(end);
        for (let bucket_index = index_start; bucket_index <= index_end; ++bucket_index) {
            const bucket_start = this.buckets[bucket_index].start;
            const bucket_end = this.buckets[bucket_index].end;
            this.buckets[bucket_index].insert(value_index, Math.max(start, bucket_start), Math.min(end, bucket_end));
        }
    }

    find(timestamp: number): Array<[number, number, number]> {
        if (!(timestamp >= this.start && timestamp <= this.end))
            return [];
        return this.buckets[this.get_bucket_index(timestamp)].find(timestamp);
    }

    find_within_range(start: number, end: number): Array<[number, number, number]> {
        start = Math.max(this.start, start);
        end = Math.min(this.end, end);
        const index_start = this.get_bucket_index(start);
        const index_end = this.get_bucket_index(end);
        const result = new Map();
        for (let bucket_index = index_start; bucket_index <= index_end; ++bucket_index) {
            for (const [value_index, b_start, b_end] of this.buckets[bucket_index].find_within_range(start, end)) {
                if (result.has(value_index)) {
                    const [r_start, r_end] = result.get(value_index);
                    result.set(value_index, [
                        Math.min(r_start, b_start),
                        Math.max(r_end, b_end)
                    ]);
                } else {
                    result.set(value_index, [b_start, b_end]);
                }
            }
        }
        // @ts-ignore
        return [...result.entries()].map(([value_index, interval]) => [value_index, ...interval]);
    }

    private get_bucket_index(timestamp: number): number {
        return Math.floor((timestamp - this.start) / this.step_size);
    }
}

File: Backend/src/modules/live_data_processor/material/interval_bucket.rs
use std::collections::HashMap;

pub trait UniqueBucketId {
    fn get_unique_bucket_id(&self) -> u64;
}

#[derive(Debug)]
struct Bucket {
    pub start: i64,
    pub end: i64,
    content: HashMap<u64, Vec<(i64, i64)>>,
}

impl Bucket {
    pub fn new(start: i64, end: i64) -> Self {
        Bucket {
            start,
            end,
            content: HashMap::new(),
        }
    }

    pub fn insert(&mut self, value_index: u64, start: i64, end: i64) {
        let content = self.content.entry(value_index).or_insert_with(Vec::new);
        content.push((start, end));
    }

    pub fn find(&self, ts: i64) -> Vec<(u64, i64, i64)> {
        self.content.iter().filter_map(|(unique_id, intervals)| intervals.iter()
            .find(|(start, end)| *start <= ts && *end >= ts).map(|(start, end)| (*unique_id, *start, *end))).collect()
    }

    pub fn find_within_range(&self, start: i64, end: i64) -> Vec<(u64, i64, i64)> {
        if start <= self.start && end >= self.end {
            self.content.keys().map(|key| (*key, start, end)).collect()
        } else if self.start <= start && end >= self.end {
            self.content.iter().filter_map(|(unique_id, intervals)| intervals.iter()
                .find(|(_, end)| *end >= start).map(|(start, end)| (*unique_id, *start, *end))).collect()
        } else if start <= self.start && self.end >= end {
            self.content.iter().filter_map(|(unique_id, intervals)| intervals.iter()
                .find(|(start, _)| *start <= end).map(|(start, end)| (*unique_id, *start, *end))).collect()
        } else {
            self.content.iter().filter_map(|(unique_id, intervals)| intervals.iter()
                .find(|(i_start, i_end)| *i_start <= end && *i_end >= start).map(|(start, end)| (*unique_id, *start, *end))).collect()
        }
    }
}

pub struct IntervalBucket<T: UniqueBucketId> {
    start: i64,
    end: i64,
    step_size: i64,
    pub value_map: HashMap<u64, T>,
    buckets: Vec<Bucket>,
}

impl<T: UniqueBucketId> IntervalBucket<T> {
    pub fn new(start: i64, end: i64, step_size: i64) -> Self {
        let num_buckets = (((end - start) as f64) / (step_size as f64)).ceil() as usize;
        let mut buckets = Vec::with_capacity(num_buckets);
        for i in 0..num_buckets {
            buckets.push(Bucket::new(start + (i as i64) * step_size, start + ((i + 1) as i64) * step_size));
        }
        IntervalBucket {
            start,
            end,
            step_size,
            buckets,
            value_map: HashMap::default(),
        }
    }

    pub fn insert(&mut self, start: i64, end: i64, value: T) {
        let start = start.max(self.start);
        let end = end.min(self.end);
        let value_index = self.value_map.entry(value.get_unique_bucket_id()).or_insert(value).get_unique_bucket_id();
        let bucket_index_start = self.get_bucket_index(start);
        let bucket_index_end = self.get_bucket_index(end);
        for bucket_index in bucket_index_start..(bucket_index_end + 1) {
            let bucket_start = self.buckets[bucket_index].start;
            let bucket_end = self.buckets[bucket_index].end;
            self.buckets[bucket_index].insert(value_index, start.max(bucket_start), end.min(bucket_end))
        }
    }

    pub fn find_unique_ids(&self, ts: i64) -> Vec<(u64, i64, i64)> {
        if !(ts >= self.start && ts <= self.end) {
            return vec![];
        }
        let bucket_index = self.get_bucket_index(ts);
        self.buckets[bucket_index].find(ts)
    }

    pub fn find_unique_ids_within_range(&self, start: i64, end: i64) -> Vec<(u64, i64, i64)> {
        let start = start.max(self.start);
        let end = end.min(self.end);
        let bucket_index_start = self.get_bucket_index(start);
        let bucket_index_end = self.get_bucket_index(end);
        let mut result = HashMap::new();
        for bucket_index in bucket_index_start..(bucket_index_end + 1) {
            for (unique_id, start, end) in self.buckets[bucket_index].find_within_range(start, end) {
                let entry = result.entry(unique_id).or_insert((start, end));
                entry.0 = entry.0.min(start);
                entry.1 = entry.1.max(end);
            }
        }
        result.iter().fold(Vec::with_capacity(result.len()), |mut acc, (unique_id, (start, end))| {
            acc.push((*unique_id, *start, *end));
            acc
        })
    }

    fn get_bucket_index(&self, ts: i64) -> usize {
        return (((ts - self.start) as f64) / (self.step_size as f64)).floor() as usize;
    }
}